### ISSUE TITLE
Latent diffusion models

### ISSUE THREAD

**Issue by amanikiruga** [sun-umn/DMPlug#2]
Hi authors, 
thank you for this great work! I was wondering if this method could easily be adapted to large latent diffusion models like stable diffusion and on diverse datasets like imagenet and laion5b. Is there any challenge you anticipate trying this? 
Additionally, did you consider trying this during your experiments? How did it go?

**Comment by HengkangWang** [comment]
Hi @amanikiruga, thanks for your interests and this good question! We tested our method with Stable Diffusion 2.1-base for super-resolution and inpainting tasks. The limited results below demonstrate reasonable performance. For simplicity, the text conditioning was set to an empty string throughout the experiments. (SD-SR4 indicates the pre-trained 4* super-resolution model from Stable Diffusion)

![Inp](https://github.com/user-attachments/assets/c81332f1-2d41-4e72-bd8b-eda26854319e)
![SR](https://github.com/user-attachments/assets/b377f67f-50a5-4471-8a3c-e57a3a90d88c)

**Comment by amanikiruga** [comment]
Amazing! Thanks for the quick response. Would you mind sharing the hyperparameters (ddim steps, lr, optimizer etc.) you used with SD2-1 and roughly how many iterations you needed? I tried with SD1-5 but didn't get such nice results. 
I assume mostly everything stays the same and you optimize the noisy Z_T latent but supervised at the pixel level through the VAE decoder? 

Much appreciated Hengkang

**Comment by HengkangWang** [comment]
You're welcome! I recall that we used 4 DDIM steps, a learning rate of 0.01 with the ADAM optimizer, and 6,000 iterations. We might also release the code with support for Stable Diffusion in the future.

**Comment by amanikiruga** [comment]
Gotcha! Was the empty string conditioning to avoid doing CFG? 

Will update if I get it to work in the meantime

**Comment by HengkangWang** [comment]
CFG scale is set as default, i.e., 9.0. Good luck!

**Comment by amanikiruga** [comment]
Hey Hengkang, 

No luck as of yet. 
Here is my code: 

```py
import argparse, os, yaml
import torch
import numpy as np
from PIL import Image
from skimage.metrics import peak_signal_noise_ratio, structural_similarity
import matplotlib.pyplot as plt
from util.img_utils import Blurkernel, clear_color, generate_tilt_map, mask_generator
from guided_diffusion.measurements import get_noise, get_operator
from guided_diffusion.unet import create_model
from diffusers import DDIMScheduler, StableDiffusionPipeline
import shutil
import lpips
from tqdm import tqdm


def load_yaml(file_path: str) -> dict:
    with open(file_path) as f:
        config = yaml.load(f, Loader=yaml.FullLoader)
    return config

def dmplug_sd(model, scheduler, logdir, img='00000', eta=0, lr=1e-2, dataset='celeba',img_model_config=None,task_config=None,device='cuda', cfg_guidance_scale=9.0):
    dtype = torch.float32
    gt_img_path = './data/{}/{}.png'.format(dataset,img)
    gt_img = Image.open(gt_img_path).convert("RGB")
    shutil.copy(gt_img_path, os.path.join(logdir, 'gt.png'))
    ref_numpy = np.array(gt_img) / 255.0
    x = ref_numpy * 2 - 1
    x = x.transpose(2, 0, 1)
    ref_img = torch.Tensor(x).to(dtype).to(device).unsqueeze(0)
    ref_img.requires_grad = False
    # Prepare Operator and noise
    measure_config = task_config['measurement']
    operator = get_operator(device=device, **measure_config['operator'])
    noiser = get_noise(**measure_config['noise'])

    if measure_config['operator']['name'] == 'inpainting':
        mask_gen = mask_generator(
            **measure_config['mask_opt']
        )
        mask = mask_gen(ref_img)
        mask = mask[:, 0, :, :].unsqueeze(dim=0)
        # Forward measurement model (Ax + n)
        y = operator.forward(ref_img, mask=mask)
        y_n = noiser(y)
    else:
        # Forward measurement model (Ax + n)
        y = operator.forward(ref_img)
        y_n = noiser(y)

    y_n.requires_grad = False
    plt.imsave(os.path.join(logdir, 'measurement.png'), clear_color(y_n))

    
    # prepare prompts 
    prompts = [""]
    # prompts = ["Cocker spaniel standing on a field of grass"]
    prompt_ids = pipeline.tokenizer(
                    prompts,
                    return_tensors="pt",
                    padding="max_length",
                    truncation=True,
                    max_length=pipeline.tokenizer.model_max_length,
                ).input_ids.to(device)   
    prompt_embeds = pipeline.text_encoder(prompt_ids)[0].to(device)
    
    # negative prompts in this case are also just the empty text so we can set them to prompt_embeds 
    # neg_prompt_embeds = prompt_embeds
    neg_prompt_ids = pipeline.tokenizer(
                    [""],
                    return_tensors="pt",
                    padding="max_length",
                    truncation=True,
                    max_length=pipeline.tokenizer.model_max_length,
                ).input_ids.to(device)

    neg_prompt_embeds = pipeline.text_encoder(neg_prompt_ids)[0].to(device)

    
    # DMPlug
    Z = torch.randn((1, 4, 64, 64), device=device, dtype=dtype, requires_grad=True)
    criterion = torch.nn.MSELoss().to(device)
    params_group1 = {'params': Z, 'lr': lr}
    optimizer = torch.optim.Adam([params_group1])

    # epochs = 5000 # SR, inpainting: 5,000, nonlinear deblurring: 10,000
    epochs = 50000 # SR, inpainting: 5,000, nonlinear deblurring: 10,000
    psnrs = []
    ssims = []
    losses = []
    lpipss = []
    best_img = None
    loss_fn_alex = lpips.LPIPS(net='alex').to(device)
    for iterator in tqdm(range(epochs), desc='Training Progress'):
        model.eval()
        optimizer.zero_grad()

        for i, tt in enumerate(pipeline.scheduler.timesteps):
            # t = (torch.ones(1) * tt).cuda()
            t = (torch.ones(1).cuda() * tt)
            if i == 0:
                noise_pred_uncond = model(Z, t, neg_prompt_embeds).sample
                noise_pred_cond = model(Z, t, prompt_embeds).sample
            else:
                noise_pred_uncond = model(x_t, t, neg_prompt_embeds).sample
                noise_pred_cond = model(x_t, t, prompt_embeds).sample
            grad = (noise_pred_cond - noise_pred_uncond)
            noise_pred = noise_pred_uncond + cfg_guidance_scale * grad                
            
            if i == 0:
                x_t = scheduler.step(noise_pred, tt, Z, return_dict=True, use_clipped_model_output=True, eta=eta).prev_sample
            else:
                x_t = scheduler.step(noise_pred, tt, x_t, return_dict=True, use_clipped_model_output=True, eta=eta).prev_sample
        ims = pipeline.vae.decode(x_t.to(pipeline.vae.dtype) / 0.18215).sample
        output = torch.clamp(ims, -1, 1)
        if measure_config['operator']['name'] == 'inpainting':
            loss = criterion(operator.forward(output, mask=mask), y_n)
        else:
            loss = criterion(operator.forward(output), y_n)
        loss.backward()
        optimizer.step()
        losses.append(loss.item())

        with torch.no_grad():
            output_numpy = output.detach().cpu().squeeze().numpy()
            output_numpy = (output_numpy + 1) / 2
            output_numpy = np.transpose(output_numpy, (1, 2, 0))
            # calculate psnr
            tmp_psnr = peak_signal_noise_ratio(ref_numpy, output_numpy)
            psnrs.append(tmp_psnr)
            # calculate ssim
            tmp_ssim = structural_similarity(ref_numpy, output_numpy, channel_axis=2, data_range=1)
            ssims.append(tmp_ssim)
            # calculate lpips
            rec_img_torch = torch.from_numpy(output_numpy).permute(2, 0, 1).unsqueeze(0).float().to(device)
            gt_img_torch = torch.from_numpy(ref_numpy).permute(2, 0, 1).unsqueeze(0).float().to(device)
            rec_img_torch = rec_img_torch * 2 - 1
            gt_img_torch = gt_img_torch * 2 - 1
            lpips_alex = loss_fn_alex(gt_img_torch, rec_img_torch).item()
            lpipss.append(lpips_alex)
            
            if len(psnrs) == 1 or (len(psnrs) > 1 and tmp_psnr > np.max(psnrs[:-1])):
                best_img = output_numpy
                
            if iterator % 10 == 0:
                print(f'Epoch {iterator}: Loss = {loss.item()}, PSNR = {tmp_psnr}, SSIM = {tmp_ssim}, LPIPS = {lpips_alex}')
                if best_img is not None:
                    plt.imsave(os.path.join(logdir, "rec_img.png"), best_img)
                    
    plt.imsave(os.path.join(logdir, "rec_img.png"), best_img)

    plt.plot(np.array(losses), label='all')
    plt.legend()
    plt.savefig(os.path.join(logdir, 'loss.png'))
    plt.close()

    plt.plot(np.array(psnrs))
    plt.title('Max PSNR: {}'.format(np.max(np.array(psnrs))))
    plt.savefig(os.path.join(logdir, 'psnr.png'))
    plt.close()

    psnr_res = np.max(psnrs)
    ssim_res = np.max(ssims)
    lpips_res = np.min(lpipss)
    print('PSNR: {}, SSIM: {}, LPIPS: {}'.format(psnr_res, ssim_res, lpips_res))

def get_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-e",
        "--eta",
        type=float,
        nargs="?",
        help="eta for ddim sampling (0.0 yields deterministic sampling)",
        default=0.0
    )
    parser.add_argument(
        "-l",
        "--logdir",
        type=str,
        nargs="?",
        help="logdir",
        default="./results"
    )
    parser.add_argument(
        "--dataset",
        type=str,
        nargs="?",
        help="dataset",
        default="stable_diffusion"
    )
    parser.add_argument(
        "-c",
        "--custom_steps",
        type=int,
        nargs="?",
        help="number of steps for ddim and fast sampling",
        default=4
    )
    parser.add_argument(
        "--lr",
        type=float,
        nargs="?",
        help="lr of z",
        default=0.01
    )
    parser.add_argument(
        "--task",
        type=str,
        nargs="?",
        help="super_resolution,inpainting,nonlinear_deblur,inpainting_box",
        default='inpainting_box'
    )
    parser.add_argument(
        "--img",
        type=int,
        nargs="?",
        help="image id",
        default=0
    )
    parser.add_argument(
        "--cfg_guidance_scale",
        type=int,
        nargs="?",
        help="cfg_guidance_scale",
        default=9.0
    )
    return parser
def torch_seed(seed=0):
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    torch.cuda.manual_seed_all(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)

if __name__ == "__main__":
    torch_seed(123)
    # Load configurations
    parser = get_parser()
    device = torch.device("cuda")
    opt, unknown = parser.parse_known_args()
    task_config = 'configs/tasks/{}_config.yaml'.format(opt.task)
    
    dtype = torch.float32
    pretrained_model_id = "stabilityai/stable-diffusion-2-1"
    pipeline = StableDiffusionPipeline.from_pretrained(pretrained_model_id, torch_dtype=dtype)

    pipeline.vae.requires_grad_(False)
    pipeline.text_encoder.requires_grad_(False)
    pipeline.unet.requires_grad_(False)
    pipeline.vae.eval()
    pipeline.text_encoder.eval()
    pipeline.unet.eval()
    
    # put to gpu 
    pipeline.vae.to(device, dtype=dtype)
    pipeline.text_encoder.to(device, dtype=dtype)
    pipeline.unet.to(device, dtype=dtype)
    
    model = pipeline.unet 

    model = model.to(device)
    model.eval()
    task_config = load_yaml(task_config)
    
    # Define the DDIM scheduler
    pipeline.scheduler = DDIMScheduler.from_config(pipeline.scheduler.config)
    pipeline.scheduler.set_timesteps(opt.custom_steps)
    
    img = str(opt.img).zfill(5)
    logdir = os.path.join(opt.logdir, opt.task, opt.dataset, img)
    os.makedirs(logdir,exist_ok=True)
    
    # DMPlug
    dmplug_sd(model, pipeline.scheduler, logdir, img=img, eta=opt.eta, lr=opt.lr, dataset=opt.dataset, task_config = task_config, device=device, cfg_guidance_scale=opt.cfg_guidance_scale)

```
    
and I also use empty prompt but with cfg. 
   
attached is also the measurement and ground truth I'm using. 
![gt](https://github.com/user-attachments/assets/603b5d96-10ab-4239-b5e5-ef26c962bb88)
![measurement](https://github.com/user-attachments/assets/8ff771af-1e0f-4606-b946-3363d19cec17)
![rec_img](https://github.com/user-attachments/assets/1c31fc96-2b0b-42dc-bf4e-3ef6d55982fe)


but the result is clearly invalid 

Thanks in advance

**Comment by amanikiruga** [comment]
ended up with 
![image](https://github.com/user-attachments/assets/ac926087-3f5c-4001-bbb4-89fb76e7027e)

but still not quite right?

**Comment by HengkangWang** [comment]
Hi @amanikiruga, thanks for trying! We may provide additional information in our upcoming publications.
